File: csharp/DesiArchitect.DesignPatterns/04_Decorator/After/ConvenienceFeeDecorator.cs
namespace DesiArchitect.DesignPatterns.Decorator.After;

public class ConvenienceFeeDecorator : IBill
{
    private readonly IBill _inner;
    private readonly decimal _fee = 5.0m;

    public ConvenienceFeeDecorator(IBill inner) => _inner = inner;

    public decimal Total()
    {
        var subtotal = _inner.Total();
        var result = subtotal + _fee;
        Console.WriteLine($"  🎫  + Convenience fee: ₹{_fee} -> ₹{result}");
        return result;
    }
}

File: typescript/src/04-decorator/after/convenience-fee-decorator.ts
import type { IBill } from './ibill.js';

export class ConvenienceFeeDecorator implements IBill {
  private readonly fee = 5;

  constructor(private readonly inner: IBill) {}

  total(): number {
    const subtotal = this.inner.total();
    const result = subtotal + this.fee;
    console.log(`  🎫  + Convenience fee: ₹${this.fee} -> ₹${result}`);
    return result;
  }
}